### ISSUE TITLE
Metabot cannot share attachments to Slack

### ISSUE THREAD

**Issue by buraks** [metabase/metabase#4056]
![image](https://cloud.githubusercontent.com/assets/2693237/21639744/8e7f38ae-d269-11e6-95d2-1945c0cc10e9.png)

Running metabase v0.21.1 with the jar file I downloaded inside a docker container. Everything works fine except Slack integration. I have created the channel `metabase_files`. Metabot can message when asked `list` or `help` fine, however `show {#}` fails with this error on the logs:

```
01-04 10:25:45 ERROR metabase.metabot :: [
"pulse.render$render_to_png.invokeStatic(render.clj:212)"
"pulse.render$render_to_png.invoke(render.clj:188)"
"pulse.render$render_html_to_png.invokeStatic(render.clj:222)"
"pulse.render$render_html_to_png.invoke(render.clj:215)"
"pulse.render$render_pulse_card_to_png.invokeStatic(render.clj:464)"
"pulse.render$render_pulse_card_to_png.invoke(render.clj:460)"
"pulse$create_and_upload_slack_attachments_BANG_$iter__29936__29940$fn__29941$fn__29942.invoke(pulse.clj:49)"
"pulse$create_and_upload_slack_attachments_BANG_$iter__29936__29940$fn__29941.invoke(pulse.clj:48)"
"pulse$create_and_upload_slack_attachments_BANG_.invokeStatic(pulse.clj:48)"
"pulse$create_and_upload_slack_attachments_BANG_.invoke(pulse.clj:44)"
"metabot$show$fn__30071.invoke(metabot.clj:129)"
]
```

Here is the Dockerfile:
```
FROM java:openjdk-8-jre-alpine

ENV FC_LANG en-US \
    LC_CTYPE en_US.UTF-8

COPY . /app
WORKDIR /app

CMD ["/app/bin/start"]
```

**Comment by camsaul** [comment]
@buraks is there anything else in the logs? Just the stacktrace?

**Comment by buraks** [comment]
Hi @camsaul this is what happens.

- I send a message on Slack `metabot show 4`
- On the logs I see
```
Jan 10 13:13:43 DEBUG metabase.query-processor :: 
QUERY: 😎
{:database 2,
:type "native",
:native
{:query
(MY_QUERY, hidden),
:info {:executed-by 3, :card-id nil, :uuid "e14854a5-761a-4d31-8cb9-6bf8f8baf7cd", :query-hash 78211612, :query-type "native"}}

Jan 10 13:13:43 DEBUG metabase.query-processor :: NATIVE FORM: 😳
{:query
(MY_QUERY, hidden)}

Jan 10 13:13:48 ERROR metabase.metabot :: ["pulse.render$render_to_png.invokeStatic(render.clj:212)" "pulse.render$render_to_png.invoke(render.clj:188)" "pulse.render$render_html_to_png.invokeStatic(render.clj:222)" "pulse.render$render_html_to_png.invoke(render.clj:215)" "pulse.render$render_pulse_card_to_png.invokeStatic(render.clj:464)" "pulse.render$render_pulse_card_to_png.invoke(render.clj:460)" "pulse$create_and_upload_slack_attachments_BANG_$iter__29936__29940$fn__29941$fn__29942.invoke(pulse.clj:49)" "pulse$create_and_upload_slack_attachments_BANG_$iter__29936__29940$fn__29941.invoke(pulse.clj:48)" "pulse$create_and_upload_slack_attachments_BANG_.invokeStatic(pulse.clj:48)" "pulse$create_and_upload_slack_attachments_BANG_.invoke(pulse.clj:44)" "metabot$show$fn__30071.invoke(metabot.clj:129)"]
```

- and metabot replies on Slach with `Uh oh! :cry:`

**Comment by buraks** [comment]
Then afterwards metabot stops responding on Slack for a while. Maybe the service gets killed and something restarts it after some time?

**Comment by buraks** [comment]
I just tried on v22.0, I still experience the same issue.

Is there something missing in the jar?

**Comment by camsaul** [comment]
If MetaBot stops replying for a while it sounds like the thread is getting killed and then restarted. Is there anything strange about the card you're trying to show? Is it a particularly large Card? 

Can you try a different Card and see if that works?

**Comment by buraks** [comment]
I have tried many other cards, none of them is particularly large. They load and render under 1 second on web. It fails on createLayout() here I believe, https://github.com/metabase/metabase/blob/master/src/metabase/pulse/render.clj#L212

Can we somehow make it print the java stack trace?

Also, I have tried getting pulse send me an email, it works fine and the images are there. If the same code renders the graphics on the email, it shouldn't be about image rendering.

**Comment by camsaul** [comment]
Looks like it's already printing the stacktrace

**Comment by camsaul** [comment]
For whatever reason it's not printing the exception message tho so I can make some tweaks to do that which should help us figure out exactly what the problem is

**Comment by miry** [comment]
According the manual for bot https://api.slack.com/bot-users need to specify the param `as_user` to `true`.

```
To post more complex messages as a bot user, clients can use the Web API method chat.postMessage. Set as_user to true to send messages as your bot with its username and profile image.
```

I see next error: http://prntscr.com/e8ytwd

```
Ok, just a second...

Uh oh! :cry:
>Slack API error: user_is_bot
```

**Comment by listentojohan** [comment]
I think I encountered the same problem. I just had metabot show a question. Then I updated the question a bit and made a pulse. Now neither "metabot list" or "metabot show <question>" works.

**Comment by Leen15** [comment]
Same problem here after upgraded the container to last version 0.24.2:

`
Jun 26 12:20:59 ERROR metabase.metabot :: ["pulse.render$render_to_png.invokeStatic(render.clj:215)" "pulse.render$render_to_png.invoke(render.clj:190)" "pulse.render$render_html_to_png.invokeStatic(render.clj:224)" "pulse.render$render_html_to_png.invoke(render.clj:217)" "pulse.render$render_pulse_card_to_png.invokeStatic(render.clj:467)" "pulse.render$render_pulse_card_to_png.invoke(render.clj:464)" "pulse$create_and_upload_slack_attachments_BANG_$iter__37379__37383$fn__37384$fn__37385.invoke(pulse.clj:48)" "pulse$create_and_upload_slack_attachments_BANG_$iter__37379__37383$fn__37384.invoke(pulse.clj:47)" "pulse$create_and_upload_slack_attachments_BANG_.invokeStatic(pulse.clj:47)" "pulse$create_and_upload_slack_attachments_BANG_.invoke(pulse.clj:43)" "metabot$show$fn__37576.invoke(metabot.clj:132)"]
`

Any idea on how to solve this?

**Comment by mazameli** [comment]
Thread on Discourse related to the Metabot permissions issue: http://discourse.metabase.com/t/slack-bot-permissions-issue/1542/5

**Comment by Leen15** [comment]
@mazameli I gave to metabot full access to all databases. For me the error is `"Can't create output stream!"` so the metabot executes the query but cannot create the png to send in slack.

**Comment by andruby** [comment]
I've also not been able to get metabot to "create an output stream"

![slack_-_silverfin](https://user-images.githubusercontent.com/10639/29278707-d475f068-8115-11e7-86ab-03f702e7265f.png)

(this is inside the channel `#metabase_files`. Note that I gave the bot the username `metabase_bot` iso `metabot`, that's the only difference from the install docs. Metabot has full access to all queries.

**Comment by Leen15** [comment]
Any news here? 
I'm on version v0.27.2.
the #metabase_files channel is public and metabot is invited inside.
I tried also while configuring a new pulse. The email pulse works, the slack pulse no.

The error is always the same:
`4/1/2018 15:12:3101-04 15:12:31 [1mERROR metabase.middleware[0m :: [31mPOST /api/pulse/test 500 (644 ms) (8 DB calls)[0m
4/1/2018 15:12:31{:message "Can't create output stream!",
4/1/2018 15:12:31 :stacktrace
4/1/2018 15:12:31 ["pulse.render$render_to_png.invokeStatic(render.clj:230)"
4/1/2018 15:12:31  "pulse.render$render_to_png.invoke(render.clj:205)"
4/1/2018 15:12:31  "pulse.render$render_html_to_png.invokeStatic(render.clj:239)"
4/1/2018 15:12:31  "pulse.render$render_html_to_png.invoke(render.clj:232)"
4/1/2018 15:12:31  "pulse.render$render_pulse_card_to_png.invokeStatic(render.clj:529)"
4/1/2018 15:12:31  "pulse.render$render_pulse_card_to_png.invoke(render.clj:526)"
4/1/2018 15:12:31  "pulse$create_slack_attachment_data$iter__31131__31135$fn__31136$fn__31137$fn__31140.invoke(pulse.clj:68)"
4/1/2018 15:12:31  "pulse$create_and_upload_slack_attachments_BANG_$iter__31154__31158$fn__31159$fn__31160.invoke(pulse.clj:79)"
4/1/2018 15:12:31  "pulse$create_and_upload_slack_attachments_BANG_$iter__31154__31158$fn__31159.invoke(pulse.clj:78)"
4/1/2018 15:12:31  "pulse$create_and_upload_slack_attachments_BANG_.invokeStatic(pulse.clj:77)"
4/1/2018 15:12:31  "pulse$create_and_upload_slack_attachments_BANG_.invoke(pulse.clj:74)"
4/1/2018 15:12:31  "pulse$fn__31242.invokeStatic(pulse.clj:200)"
4/1/2018 15:12:31  "pulse$fn__31242.invoke(pulse.clj:198)"
4/1/2018 15:12:31  "pulse$send_notifications_BANG_.invokeStatic(pulse.clj:213)"
4/1/2018 15:12:31  "pulse$send_notifications_BANG_.invoke(pulse.clj:211)"
4/1/2018 15:12:31  "pulse$send_pulse_BANG_.invokeStatic(pulse.clj:239)"
4/1/2018 15:12:31  "pulse$send_pulse_BANG_.doInvoke(pulse.clj:230)"
4/1/2018 15:12:31  "api.pulse$fn__31412$fn__31415.invoke(pulse.clj:152)"
4/1/2018 15:12:31  "api.common.internal$do_with_caught_api_exceptions.invokeStatic(internal.clj:248)"
4/1/2018 15:12:31  "api.common.internal$do_with_caught_api_exceptions.invoke(internal.clj:243)"
4/1/2018 15:12:31  "api.pulse$fn__31412.invokeStatic(pulse.clj:144)"
4/1/2018 15:12:31  "api.pulse$fn__31412.invoke(pulse.clj:144)"
4/1/2018 15:12:31  "middleware$enforce_authentication$fn__31785.invoke(middleware.clj:122)"
4/1/2018 15:12:31  "api.routes$fn__45022.invokeStatic(routes.clj:79)"
4/1/2018 15:12:31  "api.routes$fn__45022.invoke(routes.clj:79)"
4/1/2018 15:12:31  "routes$fn__45721$fn__45722.doInvoke(routes.clj:101)"
4/1/2018 15:12:31  "routes$fn__45721.invokeStatic(routes.clj:97)"
4/1/2018 15:12:31  "routes$fn__45721.invoke(routes.clj:97)"
4/1/2018 15:12:31  "middleware$log_api_call$fn__31884$fn__31886.invoke(middleware.clj:330)"
4/1/2018 15:12:31  "middleware$log_api_call$fn__31884.invoke(middleware.clj:329)"
4/1/2018 15:12:31  "middleware$add_security_headers$fn__31834.invoke(middleware.clj:245)"
4/1/2018 15:12:31  "middleware$bind_current_user$fn__31789.invoke(middleware.clj:142)"
4/1/2018 15:12:31  "middleware$maybe_set_site_url$fn__31838.invoke(middleware.clj:268)"]}`

**Comment by Fogli** [comment]
hi, i'm using version 0.29.3 and i'm having the same issue. 
The answer of the bot to the show ID_QUESTION command is 
"you don't have permission to do that". I have checked all the permission for the bot in the admin panel but he has access to all the db. Any news or trick to make it works? (Not a developer than I cannot show you a full log with errors). Thanks ad good job guys. The product is awesome!

**Comment by lp-lima** [comment]
> hi, i'm using version 0.29.3 and i'm having the same issue.
> The answer of the bot to the show ID_QUESTION command is
> "you don't have permission to do that". I have checked all the permission for the bot in the admin panel but he has access to all the db. Any news or trick to make it works? (Not a developer than I cannot show you a full log with errors). Thanks ad good job guys. The product is awesome!

Check this [issue](https://github.com/metabase/metabase/issues/7660), it might help you in that matter,

**Comment by flamber** [comment]
There was a lot of fixes in 0.36.x (latest release is currently 0.41.1): https://github.com/metabase/metabase/releases/latest

This issue hasn't been reported for several years, so closing.
And the entire Slack implementation will be rewritten in the future - see #12383

If you are still experiencing this issue, then please include "Diagnostic Info" from Admin > Troubleshooting, and the full stacktrace from Admin > Troubleshooting > Logs, and any other helpful information (like which Slack commands you're using etc).
